test/parse.jl

```diff
@@ -58,24 +58,50 @@ end
 
 end
 @testset "validating commands" begin
-    @test well_formed("daw") == true
-    @test well_formed("aw") == false
-    @test well_formed("5d5w5") == false
-    @test well_formed("dd") == true
-    @test well_formed("yy") == true
-    @test well_formed("5dd") == true
-    @test well_formed("yy") == true
-    @test well_formed("yd") == false
-    @test well_formed("cd") == false
-
-    @test well_formed("h") == true
-    @test well_formed("l") == true
-    @test well_formed("5l") == true
-    @test well_formed("4h") == true
-
-    @test well_formed("10dd") == true
+    well_formed_cmds = (
+        "daw", "dd", "yy", "5dd", "yy", "h", "l", "5l", "4h", "10dd", "x",
+        "ri", "ra", "r0", "r7", "5fd", "5Fd", "ciw", "dw", "2dw", "j", "k",
+        "I", "O", "o", "10D", "20x", "2dfe", "2cFk", "2dF0"
+    )
+    poorly_formed_cmds = (
+        "aw", "5d5w5", "yd", "cd"
+    )
+    @testset "Well formed commands" begin
+        for cmd in well_formed_cmds
+            @test well_formed(cmd) || @show cmd
+        end
+        for cmd in poorly_formed_cmds
+            @test !well_formed(cmd) || @show cmd
+        end
+    end
+    @testset "Partially well formed commands" begin
+        # Well formed commands should be partially well formed
+        #  when sliced from their start to any point in the middle.
+        for cmd in well_formed_cmds
+            for cmd_end in 1:length(cmd)
+                cmd_stub = cmd[1:cmd_end]
+                @test partial_well_formed(cmd_stub) || @show cmd_stub
+            end
+        end
+        for cmd in poorly_formed_cmds
+            @test !partial_well_formed(cmd) || @show cmd
+        end
+        # TODO: Remove these when the commands are implemented:
+        @test !partial_well_formed("u")
```

Comment:
Can you clarify what is to be removed? Is it the negations using `!` when a command has been implemented?

In that case, I would prefer to use the correct test expression with `@test_skip` or `@test_broken`.

```suggestion
        @test_skip partial_well_formed("u")
```

Reply:
Right; I mean that the checks here should be inverted (`!partial_well_formed -> partial_well_formed`) once the commands are implemented (like `u` for undo). But since typing `u` will not do anything right now, and is not part of any well formed command, it seems like clearing the stack is the right thing to do, otherwise you would have to hit escape manually.

I think `@test` is still what we want to use, because we want to treat `"u"` as _not_ well formed. i.e., we want the stack to be cleared on unimplemented commands, so that the user does not need to hit the escape key.

`@test_broken` should technically work, but I think it's _semantically_ improper, because we want to treat unimplemented commands as not well formed. I think `@test_skip` wouldn't perform the test at all?

What do you think?

Reply:
Ah, I understand now. I saw this as a "not implemented yet" test list, instead of code which is specifically invalidating keys that are not in the feature set yet.

And certainly the behavior you mention is correct - clearing the stack when a user hits `u` or any other un-implemented key.

Lets stick with `@test` then, semantically what we want for testing the package as it is now.

Reply:
Sounds good to me!